pkg/registry/apis/provisioning/jobs/driver.go

```diff
@@ -114,27 +126,39 @@ func (d *jobDriver) Run(ctx context.Context) {
 				logger.Error("failed to cleanup jobs", "error", err)
 			}
 		case <-jobTicker.C:
-			d.startDriving(ctx)
+			d.drive(ctx)
 		case <-d.store.InsertNotifications():
-			d.startDriving(ctx)
+			d.drive(ctx)
 		}
 	}
 }
 
-func (d *jobDriver) startDriving(ctx context.Context) {
-	timeoutCtx, cancel := context.WithTimeout(ctx, d.timeout)
-	defer cancel()
-	for timeoutCtx.Err() == nil {
-		if err := d.drive(timeoutCtx); err != nil {
-			if !errors.Is(err, context.Canceled) && !errors.Is(err, ErrNoJobs) {
+// The timing logic here is a bit odd
+// A new thread will be started:
+//   - anytime a new job is added
+//   - periodically jobInterval (eg 30s)
+//
+// Each thread (drive) runs until there are no jobs or has a system error
+func (d *jobDriver) drive(ctx context.Context) {
+	start := time.Now()
+	for {
+		err := d.claimAndProcessOneJob(ctx)
+		if err != nil {
+			if !errors.Is(err, ErrNoJobs) {
 				logging.FromContext(ctx).Error("failed to drive jobs", "error", err)
 			}
-			break
+			return
+		}
+
+		// A new thread will start at this interval,
+		// so lets avoid having too many threads running
+		if time.Since(start) > (d.jobInterval * 2) {
```

Comment:
I would stop picking up jobs close to the `d.jobInterval`, so that we let the next "tick" happen. 

```suggestion
		if time.Since(start) > (d.jobInterval - 5 * time.Second) {
```

(with a better math)

Reply:
it should be fine to pick until NoJobs or one errors. No need to measure the time.